### PR TITLE
Config merge tool

### PR DESCRIPTION
**Pull request by vladak**
not exactly a merge however it does what we need.

### CODE REVIEW

File: tools/ConfigMerge [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+            echo "Uknown option \"$opt\""
+	    Usage
+	    exit 1
+        ;;
+    esac
+    optidx=`expr $optidx + 1`
+done
+
+Info
+StdInvocation $*
```
Comment (tulinkry) [inline tools/ConfigMerge]: does this work when the file name contains white space character?

File: tools/ConfigMerge [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    echo "Maybe try --help option?"
+    exit 3
+fi
+
+SetupInstanceConfiguration
+
+optidx=1
+while [ $optidx -lt $# ]
+do
+    # opt=$( eval echo \${${optidx}} )
```
Comment (tulinkry) [inline tools/ConfigMerge]: not necessary comment

File: tools/ConfigMerge [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+}
+
+StdInvocation()
+{
+    CommonInvocation \
+    -- "${@}"
+}
+
+
+OneArgument()
```
Comment (tulinkry) [inline tools/ConfigMerge]: unused function

File: tools/ConfigMerge [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+{
+    if [ $# -gt 0 ]
+    then
+            echo "$1"
+    fi
+}
+
+ExpectOption()
```
Comment (tulinkry) [inline tools/ConfigMerge]: unused function

File: src/org/opensolaris/opengrok/configuration/ConfigMerge.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+ *
+ * @author Vladimir Kotal
+ */
+public class ConfigMerge {
+
+    private static final String name = "ConfigMerge";
+
+    public static void main(String[] argv) {
+
+        String type = null;
```
Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/ConfigMerge.java]: fixed

File: src/org/opensolaris/opengrok/configuration/ConfigMerge.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+        String text = null;
```
Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/ConfigMerge.java]: fixed

File: src/org/opensolaris/opengrok/configuration/ConfigMerge.java [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+        String className = null;
```
Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/ConfigMerge.java]: fixed

File: src/org/opensolaris/opengrok/configuration/ConfigMerge.java [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        boolean check = false;
```
Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/ConfigMerge.java]: ditto

File: src/org/opensolaris/opengrok/configuration/ConfigMerge.java [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+            System.err.println("cannot read file " + argv[optind + 1]);
+            System.exit(1);
+        }
+
+        // Basic strategy: take all non-static/transient fields that have a setter
+        // from cfgNew that are not of default value and set them to cfgBase.
+        for (Field field : cfgNew.getClass().getDeclaredFields()) {
+            String fieldName = field.getName();
+            int modifiers = field.getModifiers();
+            if (Modifier.isStatic(modifiers) || Modifier.isTransient(modifiers)) {
```
Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/ConfigMerge.java]: added

File: src/org/opensolaris/opengrok/configuration/ConfigMerge.java [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+                if (obj == null || obj.equals(getter.invoke(cfgDefault))) {
+                    continue;
+                }
+            } catch (Exception ex) {
+                System.err.println("failed to invoke getter for " + fieldName + ": " + ex);
+                System.exit(1);
+            }
+
+            try {
+                // System.err.println("invoking for " + fieldName);
```
Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/ConfigMerge.java]: removed

File: src/org/opensolaris/opengrok/configuration/Configuration.java [hunk 11]
```diff
@@ -293,7 +291,7 @@ public void setCommandTimeout(int command_timeout) throws IllegalArgumentExcepti
             throw new IllegalArgumentException(
                     String.format(NEGATIVE_NUMBER_ERROR, "command_timeout", command_timeout));
```
Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/Configuration.java]: renamed

File: tools/ConfigMerge (hunk 3 above)
Comment (vladak) [inline tools/ConfigMerge]: removed

File: tools/ConfigMerge (hunk 4 above)
Comment (vladak) [inline tools/ConfigMerge]: removed

File: tools/ConfigMerge (hunk 2 above)
Comment (vladak) [inline tools/ConfigMerge]: removed

File: src/org/opensolaris/opengrok/configuration/ConfigMerge.java [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+
+            Method getter = desc.getReadMethod();
+            if (getter == null) {
+                System.err.println("no getter for '" + fieldName + "'");
+                System.exit(1);
+            }
+
+            try {
+                Object obj = getter.invoke(cfgNew);
+                if (obj == null || obj.equals(getter.invoke(cfgDefault))) {
```
Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/ConfigMerge.java]: ok, removed the null check

File: tools/ConfigMerge (hunk 1 above)
Comment (vladak) [inline tools/ConfigMerge]: good point, changed

File: src/org/opensolaris/opengrok/configuration/Configuration.java [hunk 13]
```diff
@@ -1127,7 +1129,7 @@ public String getXMLRepresentationAsString() {
         return bos.toString();
     }
 
-    private void encodeObject(OutputStream out) {
+    public void encodeObject(OutputStream out) {
```
Comment (vladak) [inline src/org/opensolaris/opengrok/configuration/Configuration.java]: I see nothing wrong with exposing this method however am open to suggestions :-)

**Comment by tulinkry** [comment]
A basic test case would be nice to verify that this tool works fine

File: src/org/opensolaris/opengrok/configuration/Configuration.java (hunk 13 above)
Comment (tulinkry) [inline src/org/opensolaris/opengrok/configuration/Configuration.java]: Yes this is better, directly using it for System.out.

**Comment by vladak** [comment]
added simple test case. I decided not to add actual configuration XML files as this would require changing them every time Configuration representation changes.

File: platform/solaris/ips/create.sh [hunk 14]
```diff
@@ -199,6 +199,7 @@ PKG pkgsend add file platform/solaris/smf/ogindexd mode=0555 owner=root group=bi
```
Comment (tulinkry) [inline platform/solaris/ips/create.sh]: copyright

File: platform/solaris/pkgdef/prototype [hunk 15]
```diff
@@ -49,6 +49,7 @@ f manifest var/svc/manifest/application/opengrok.xml=platform/solaris/smf/opengr
 f none lib/svc/method/svc-opengrok=platform/solaris/smf/svc-opengrok 555 root bin
```
Comment (tulinkry) [inline platform/solaris/pkgdef/prototype]: copyright

**Comment by tulinkry** [comment]
change the copyrights and it's ok to merge

**Pull request merged**